File: geekbang/data-structure-and-algorithm/06-練習題.rb
class Node
  attr_accessor :next
  attr_reader :value

  def initialize(value)
    @value = value
    @next = nil
  end

end


def is_palindrome?(node)
  return true if node.nil? || node.next.nil?

  prev = nil
  slow = node
  fast = node

  while (fast != nil || fast.next != nil)
    fast = fast.next.next
    next_node = slow.next
    slow.next = prev
    prev = slow
    slow = next_node
  end

  unless fast.nil?
    slow = slow.next
  end

  while !slow.nil?
    return false if slow.value != prev.value

    slow = slow.next
    prev = prev.next
  end

  return true
end

File: geekbang/data-structure-and-algorithm/06-練習題.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */



class Solution {
  public boolean isPalindrome(ListNode head) {
    if (head == null || head.next == null) {
      return true;
    }

    ListNode prev = null;
    ListNode slow = head;
    ListNode fast = head;

    while (fast != null && fast.next!= null) {
      fast = fast.next.next;
      ListNode next = slow.next;
      slow.next = prev;
      prev = slow;
      slow = next;
    }

    if (fast != null) {
      slow = slow.next;
    }

    while (slow != null) {
      if (slow.val != prev.val) {
        return false;
      }
      slow = slow.next;
      prev = prev.next;
    }

    return true;
  }
}
